pkg/controllers/user/logging/utils/embedded.go

```diff
@@ -201,6 +210,120 @@ func RemoveEmbeddedTarget(dep rv1beta2.DeploymentInterface, sa rv1.ServiceAccoun
 	return nil
 }
 
+func UpdateEmbeddedEndpoint(podLister rv1.PodLister, serviceLister rv1.ServiceLister, clusterLoggings v3.ClusterLoggingInterface) {
+	timeout := time.After(2 * time.Minute)
+	tick := time.Tick(10 * time.Second)
+	for {
+		select {
+		case <-timeout:
+			logrus.Warn("update embedded endpoint timeout")
+			return
+		case <-tick:
+			cls, err := clusterLoggings.List(metav1.ListOptions{})
+			if err != nil {
+				logrus.Debugf("get cluterlogging failed, %v", err)
+			}
+			if len(cls.Items) != 0 {
+				cl := cls.Items[0]
```

Comment:
It merely handles the first clusterlogging item here. how do we deal with multiple clusters and multiple clusterlogging config?

Reply:
each cluster will only have one clusterlogging. And each controller will handle only one cluster, so I just use the first one here.

Reply:
No.. I think you are listing all ClusterLogging objects here. Say controller A for cluster1 and controller B for cluster1, A & B will get the same list of the ClusterLogging objects. So you need to tell which ClusterLogging object takes responsibility for this cluster.

Reply:
Yes, clusterLoggings.List will get config for all clusters.  In clusterlogging.go, `clusterloggingClient := cluster.Management.Management.ClusterLoggings("")`. Note that for each cluster there is a namespace, and the cluster scoped crd will be stored in that namespace.  If you do not specifiy the namespace, it gets all.

Reply:
the correct fix should be clusterloggingClient := cluster.Management.Management.ClusterLoggings(clusterName)

Reply:
+1

Reply:
when I add a namespace to the interface, and use the cache to list in controller, I get nothing. cluster.Management.Management.ClusterLoggings(cluster.ClusterName).Controller().Lister().List, event if I also use the Lister in the controller constructor.